internal/service/domain/tree/tree.go

```diff
@@ -62,3 +66,41 @@ func (s *TreeService) Delete(ctx context.Context, id int32) error {
 	}
 	return nil
 }
+
+func (s *TreeService) Update(ctx context.Context, id int32, tu *entities.TreeUpdate) (*entities.Tree, error) {
+	currentTree, err := s.treeRepo.GetByID(ctx, id)
+	if err != nil {
+		return nil, handleError(err)
+	}
+	// Check if the tree is readonly (imported from csv)
+	if currentTree.Readonly {
+		return nil, handleError(fmt.Errorf("tree with ID %d is readonly and cannot be updated", id))
+	}
+	fn := make([]entities.EntityFunc[entities.Tree], 0)
+	if tu.TreeClusterID != nil {
+		treeCluster, err := s.treeClusterRepo.GetByID(ctx, *tu.TreeClusterID)
+		if err != nil {
+			return nil, handleError(fmt.Errorf("failed to find TreeCluster with ID %d: %w", *tu.TreeClusterID, err))
+		}
+		fn = append(fn, tree.WithTreeCluster(treeCluster))
+	}
+	if tu.PlantingYear != 0 {
+		fn = append(fn, tree.WithPlantingYear(tu.PlantingYear))
+	}
+	if tu.Species != "" {
+		fn = append(fn, tree.WithSpecies(tu.Species))
+	}
+	if tu.Number != "" {
+		fn = append(fn, tree.WithTreeNumber(tu.Number))
+	}
+	if tu.Latitude != 0 && tu.Longitude != 0 {
+		fn = append(fn, tree.WithLatitude(tu.Latitude), tree.WithLongitude(tu.Longitude))
+	}
```

Comment:
Yes, okay... another solution, and I think a much cleaner solution, is to use some kind of validator, like https://pkg.go.dev/github.com/go-playground/validator/v10, which I already have somewhere here in this project

Reply:
I noticed that you don't use any kind of validation in your implementation for updating or creating tree clusters. I want to ask if we should follow the same approach in the tree cluster section?